File: Máximo Común Divisor/C/MCD_Por_Factores_Primos.c
#include <stdio.h>
#include <math.h>

int MCD(int a, int b);
int siguiente_primo(int n);
int es_primo(int n);

int main(){
	printf("Ingrese dos números: ");
	int a, b;
	scanf("%d %d", &a, &b);
	if (a>0 && b>0) 
		printf("El MCD(%d,%d)=%d",a, b, MCD(a,b));
	else
		printf("Números inválidos");	
	return 0;
}

int MCD(int a, int b){
	int factor=2;
	int producto = 1;
	while (factor<=a && factor<=b){
		if (a%factor==0 || b%factor==0){
			if (a%factor==0 && b%factor==0)
				producto *= factor;
			if (a%factor==0)
				a /= factor;
			if (b%factor==0)
				b /= factor;
		}
		else
			factor = siguiente_primo(factor);
	}
	return producto;
}

int es_primo(int n) {
    if (n <= 1)
        return 0;
    int encontro_divisores = 0;
    int i = 2;
    while (i <= sqrt(n) && !encontro_divisores) {
        if (n % i == 0)
            encontro_divisores = 1;
        i++;
    }
    return !encontro_divisores;
}

int siguiente_primo(int n) {
    do {
        n++;
    } while (!es_primo(n));
    return n;
}

File: Máximo Común Divisor/Java/MCD_Por_Factores_Primos.java
package MCD;

import static java.lang.Math.sqrt;
import java.util.Scanner;

public class MCD_Por_Factores_Primos {

    public void main(String[] args) {
        System.out.printf("Ingrese dos números: ");
        Scanner reader = new Scanner(System.in);
        int a = reader.nextInt();
        int b = reader.nextInt();
        if (a > 0 && b > 0) {
            System.out.printf("El MCD(%d,%d)=%d", a, b, MCD(a, b));
        } else {
            System.out.printf("Números inválidos");
        }
    }

    int MCD(int a, int b) {
        int factor = 2;
        int producto = 1;
        while (factor <= a && factor <= b) {
            if (a % factor == 0 || b % factor == 0) {
                if (a % factor == 0 && b % factor == 0) {
                    producto *= factor;
                }
                if (a % factor == 0) {
                    a /= factor;
                }
                if (b % factor == 0) {
                    b /= factor;
                }
            } else {
                factor = siguiente_primo(factor);
            }
        }
        return producto;
    }
    
    private static int siguiente_primo(int n) {
        do {
            n++;
        } while (!es_primo(n));
        return n;
    }

    private static boolean es_primo(int n) {
        if (n <= 1) {
            return false;
        }
        boolean encontro_divisores = false;
        int i = 2;
        while (i <= sqrt(n) && !encontro_divisores) {
            if (n % i == 0) {
                encontro_divisores = true;
            }
            i++;
        }
        return !encontro_divisores;
    }
}
